### PR TITLE
BUG: groupby apply on selected columns yielding scalar (GH13568)

### LINKED ISSUE #13568: Apply on selected columns of a groupby object - stopped working with 0.18.1
#### Code Sample
```python
>>> import pandas as pd
>>> import numpy as np
>>> 
>>> df = pd.DataFrame({'A' : ['foo', 'bar', 'foo', 'bar',
...                            'foo', 'bar', 'foo', 'foo'],
...                     'B' : ['one', 'one', 'two', 'three',
...                            'two', 'two', 'one', 'three'],
...                     'C' : np.random.randn(8),
...                     'D' : np.random.randn(8)})
>>> 
>>> gb = df.groupby('B')
>>> 
>>> gb[['C', 'D']].apply(lambda group: group.C.mean() - group.D.mean())

```
Works fine on 0.18.0. 
On 0.18.1 throws:
> TypeError: Series.name must be a hashable type

(some other random numbers)

### Comment
With 0.18.1 update, some piece of my code stopped working, in a quite surprising way. 

The code is easy to fix - just by removing [['C', 'D']] and applying on the whole groupby object. 

Nevertheless, its a regression in a point release and I think someone else may run into the same problem.

I am guessing that pandas is trying to generate a name for the result series based on the column list - unfortunately I cannot investigate further right now.

#### output of ``pd.show_versions()``
INSTALLED VERSIONS
------------------
commit: None
python: 3.5.1.final.0
python-bits: 64
OS: Linux
OS-release: 4.4.0-24-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8

pandas: 0.18.1
nose: None
pip: 8.1.2
setuptools: 24.0.2
Cython: None
numpy: 1.11.1
scipy: None
statsmodels: None
xarray: None
IPython: None
sphinx: None
patsy: None
dateutil: 2.5.3
pytz: 2016.4
blosc: None
bottleneck: None
tables: None
numexpr: None
matplotlib: None
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: None
boto: None
pandas_datareader: None

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
- [x] closes #13568
- [x] tests added / passed
- [x] passes `git diff upstream/master | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
@jreback Can you remember the reason that you added the `name=self.name` in several places in the  groupby/rolling/resample refactor? (see https://github.com/pydata/pandas/commit/699424027fb657192541bcd0c3d9f9b7d26f2300#diff-720d374f1a709d0075a1f0a02445cd65R3376) 
I think the fix is correct here, as AFAIU `self.name` can only be None (and that is already the default for name in Series) or a list (from `self._selection`) which raises an error as name.

**Comment by jreback** [comment]
The nested groupy-aplly (e.g. ``.rolling.groupby.mean`` does ``.rolling.groupby.apply(lambda x: mean)``, and wasn't passing things thru. However the fixing name things was in several places; it may be that this particular change was not needed.

So the path that you are changing is all scalars, so this seems ok. If nothing breaks then you are good to go!.

*though* the pass in the OP is really related to ``_selection``, so this *might* be a red herring.

**Comment by jorisvandenbossche** [comment]
> So the path that you are changing is all scalars, so this seems ok. If nothing breaks then you are good to go!.

Indeed, that was my thought as well

> though the path in the OP is really related to _selection, so this might be a red herring. 

The problem is that I am not familiar enough to know all use / corner cases of `self._selection` and `self.name`. At first sight setting  `name` equal to `_selection` in case of a selection of multiple columns (so for a DataFrameGroupBy) does not seem to make much sense to me, as you never want to use that as the name of a result

Added a comment

**Comment by jorisvandenbossche** [comment]
Actually, if I just let `name` return None instead of `_selection` (except for SeriesGroupBy, there the property is overwritten), the tests pass as well: https://github.com/jorisvandenbossche/pandas/commit/a6f5c9ec02a1c6495555d372069df052a4b6c08b

**Comment by jreback** [comment]
lgtm

File: doc/source/whatsnew/v0.18.2.txt [hunk 1]
```diff
@@ -478,6 +478,7 @@ Bug Fixes
 - Bug in ``PeriodIndex`` construction returning a ``float64`` index in some circumstances (:issue:`13067`)
 - Bug in ``.resample(..)`` with a ``PeriodIndex`` not changing its ``freq`` appropriately when empty (:issue:`13067`)
 - Bug in ``.resample(..)`` with a ``PeriodIndex`` not retaining its type or name with an empty ``DataFrame`` appropriately when empty (:issue:`13212`)
+- Bug in ``groupby(..).apply(..)`` when the passed function returns scalar values per group (:issue:`13468`).
```
Comment (jreback) [inline doc/source/whatsnew/v0.18.2.txt]: though need to rebase and move this

**Pull request merged**